arrow-flight/src/sql/server.rs

```
@@ -710,10 +710,21 @@ where
         // we wrap this stream in a `Peekable` one, which allows us to peek at
         // the first message without discarding it.
         let mut request = request.map(PeekableFlightDataStream::new);
-        let cmd = Pin::new(request.get_mut()).peek().await.unwrap().clone()?;
+        let mut stream = Pin::new(request.get_mut());
+
+        // If the stream is empty or the first item is an Err,
+        // return early with 0 records affected
+        let peeked_item = stream.peek().await.cloned();
+        let Some(cmd) = peeked_item else {
+            let result = DoPutUpdateResult { record_count: 0 };
+            let output = futures::stream::iter(vec![Ok(PutResult {
+                app_metadata: result.encode_to_vec().into(),
+            })]);
+            return Ok(Response::new(Box::pin(output)));
```

II the client is not sending a command, it probably is making a mistake, and returning an `Ok` will mask the problem rather than failing fast

Rather than return an Ok, I recommend  invoking `do_put_fallback` in this case which will error by default, and lets users override the behavior if they wish.

---

That makes sense. It looks like [`do_put_fallback`](https://github.com/apache/arrow-rs/blob/1f15130414bdfc01c8989ec95702655bf553c5c5/arrow-flight/src/sql/server.rs#L378-L382) requires providing a message of type `Any`. What do you recommend passing in for the message? 

Here are a few potential options that pop out to me:
- Make the message an Option. Unfortunately this would be user-facing modification of the callback function signature. 
- Create a new variant on the [Command enum](https://github.com/apache/arrow-rs/blob/1f15130414bdfc01c8989ec95702655bf553c5c5/arrow-flight/src/sql/mod.rs#L165-L173) like `Command::Missing` and add a corresponding protobuf message type. 
- Create a new callback for unrecoverable errors, eg. `do_put_error`, which doesn't require passing in a message. It could take some kind of error enum with different types of errors, eg `DoPutError::MissingCommand`.

---

I recommend a new callback for unrecoverable errors as that would be a backwards compatible API change that we could release in the next release (not have to wait for the next major release)

---

Implemented this.